pkg/apis/sources/v1alpha2/container_conversion_test.go

```diff
@@ -0,0 +1,228 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha2
+
+import (
+	"context"
+	"reflect"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"knative.dev/eventing/pkg/apis/sources/v1beta1"
+	"knative.dev/pkg/apis"
+	duckv1 "knative.dev/pkg/apis/duck/v1"
+)
+
+func TestContainerSourceConversionBadType(t *testing.T) {
+	good, bad := &ContainerSource{}, &dummy{}
+
+	if err := good.ConvertTo(context.Background(), bad); err == nil {
+		t.Errorf("ConvertTo() = %#v, wanted error", bad)
+	}
+
+	if err := good.ConvertFrom(context.Background(), bad); err == nil {
+		t.Errorf("ConvertFrom() = %#v, wanted error", good)
+	}
+}
+
+func TestContainerSourceConversionRoundTripUp(t *testing.T) {
+	versions := []apis.Convertible{&v1beta1.ContainerSource{}}
+
+	path, _ := apis.ParseURL("/path")
+	sink := duckv1.Destination{
+		Ref: &duckv1.KReference{
+			Kind:       "Foo",
+			Namespace:  "Bar",
+			Name:       "Baz",
+			APIVersion: "Baf",
+		},
+		URI: path,
+	}
+	sinkUri, _ := apis.ParseURL("http://example.com/path")
+
+	tests := []struct {
+		name string
+		in   *ContainerSource
+	}{{name: "empty",
+		in: &ContainerSource{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:       "ping-name",
+				Namespace:  "ping-ns",
+				Generation: 17,
+			},
+			Spec: ContainerSourceSpec{},
+			Status: ContainerSourceStatus{
+				SourceStatus: duckv1.SourceStatus{
+					Status: duckv1.Status{
+						ObservedGeneration: 1,
+						Conditions: duckv1.Conditions{{
+							Type:   "Ready",
+							Status: "True",
+						}},
+					},
+				},
+			},
+		},
+	}, {name: "simple configuration",
+		in: &ContainerSource{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:       "ping-name",
+				Namespace:  "ping-ns",
+				Generation: 17,
+			},
+			Spec: ContainerSourceSpec{
+				SourceSpec: duckv1.SourceSpec{
+					Sink: sink,
+				},
+			},
+			Status: ContainerSourceStatus{
+				SourceStatus: duckv1.SourceStatus{
+					Status: duckv1.Status{
+						ObservedGeneration: 1,
+						Conditions: duckv1.Conditions{{
+							Type:   "Ready",
+							Status: "Unknown",
+						}},
+					},
+					SinkURI: sinkUri,
+				},
+			},
+		},
+	}, {name: "full",
+		in: &ContainerSource{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:       "ping-name",
+				Namespace:  "ping-ns",
+				Generation: 17,
+			},
+			Spec: ContainerSourceSpec{
+				SourceSpec: duckv1.SourceSpec{
```

Comment:
can you add a simple PodTemplate here as well?

Reply:
Done